django/db/models/fields/composite.py

```diff
@@ -0,0 +1,190 @@
+                return type_(values)
+            return value
+
+        return resolve_value(self)
+
+    def __init__(self, instance):
+        self._instance = instance
+
+    def __repr__(self):
+        "Return a nicely formatted representation string"
+        return self.__class__.__name__ + self._repr_fmt % tuple(self)
+
+    def _asdict(self):
+        "Return a new dict which maps field names to their values."
+        return dict(zip(self._fields, self))
+
+    def __len__(self):
+        return len(self._fields)
+
+    def __iter__(self):
+        return (getattr(self._instance, name) for name in self._fields)
+
+    def __eq__(self, other):
+        if isinstance(other, CompositeType):
+            return tuple(self).__eq__(tuple(other))
+        return tuple(self).__eq__(other)
+
+    def __getnewargs__(self):
+        "Return self as a plain tuple.  Used by copy and pickle."
+        return tuple(self)
+
+    def __getitem__(self, i):
+        return getattr(self._instance, self._fields[i])
+
+    def __setitem__(self, i, v):
+        return setattr(self._instance, self._fields[i], v)
+
+
+class CompositeDeferredAttribute(DeferredAttribute):
+    def _bind(self, instance, cls):
+        if issubclass(cls, CompositeType):
+            return cls(instance)
+        vals = [getattr(instance, name) for name in self.field.component_names]
+        return cls(*vals)
+
+    def __get__(self, instance, cls=None):
+        if instance is None:
+            return self
+        data = instance.__dict__
+        field_name = self.field.attname
+        if field_name not in data:
+            data[field_name] = self._bind(instance, cls=self.field.python_type)
+        return data[field_name]
+
+    def __set__(self, instance, value):
+        for name, val in zip(self.field.component_names, value):
+            setattr(instance, name, val)
+
+
+class CompositeField(Field):
+    component_names = None
+    composite_columns = None
+    python_type = tuple
+
+    descriptor_class = CompositeDeferredAttribute
+
+    def __init__(self, *names, **kwargs):
+        kwargs["db_column"] = None
+        kwargs["editable"] = False
+        super().__init__(**kwargs)
+        self.component_names = tuple(names)
+        # this would deferred to post_init
+        self.component_fields = None
+
+    def get_attname_column(self):
+        return self.get_attname(), self.db_column
+
+    def contribute_to_class(self, cls, name, private_only=False) -> None:
+        # logger.info(f"{cls.__name__} {name} {cls._meta} {self}")
+        self.python_type = CompositeType.namedtuple(
+            f"{cls.__name__}__{name}", self.component_names
+        )
+        result = super().contribute_to_class(cls, name, private_only)
+        setattr(cls, self.attname, self.descriptor_class(self))
+
+        if self.primary_key:
+            cls._meta.pk = self
+        if self.unique or self.primary_key:
+            cls._meta.unique_together = (
```

Comment:
shouls you also consider UniqueConstraint https://docs.djangoproject.com/en/4.1/ref/models/options/#django.db.models.Options.unique_together?

Reply:
I've searched `UniqueConstraint` and `unique_together` in code base, found most inner code using the latter now.
I think the main difference is that `UniqueConstraint` have to construct a unique name across database.